src/Data/NodeResolver.php

```diff
@@ -45,6 +45,24 @@ public function __construct( AppContext $context ) {
 	 */
 	public function resolve_uri( string $uri, $extra_query_vars = '' ) {
 
+		/**
+		 * When this filter return anything other than null, it will be used as a resolved node
+		 * and the execution will be skipped.
+		 *
+		 * This is to be used in extensions to resolve their own nodes which might not use
+		 * WordPress permalink structure.
+		 *
+		 * @param null $node The node, defaults to nothing.
+		 * @param string $uri The uri being searched.
+		 * @param AppContext $content The app context.
+		 * @param WP $wp WP object.
+		 */
+		$node = apply_filters( 'graphql_pre_resolve_uri', null, $uri, $this->context, $this->wp );
```

Comment:
Nice. I think this is indeed a good addition!

Reply:
Now, the trick is handling cases where the node to return is actually null. 🤔 

Like, what if we had some private data that we might resolve like so (psuedo code): 

```php
add_filter( 'graphql_pre_resolve_uri', function( $node, $uri ) {

   $node_in_db = some_function_to_get_node();
   
   if ( this_node_is_private( $node_in_db ) ) {
     return null;
   } 
   
   return $node;

});
```

In this example, we'd be resolving the data, determining there is no data, or the data isn't safe to show to the requesting user, so we return null. 

Now, the NodeResolver will have executed _that_ and will continue to execute the remaining logic instead of exiting early. 

🤔

Reply:
@jasonbahl 

I put the hook in the top for a few reasons. Maybe very specific to my extension. 

- Currently, external URI is block.
- Homepage is blocked
- URLs return `null` if they are not match to any of the well known WP objects
- My extension URLs are customs, so there was no need to pass by the current `resolve_uri` logic first

In the scenario presented is okay to let the `resolve_uri` execution continues since the query is not type specific. It can match in a type and not in another. If you bail early, you might not get the node you need.

Reply:
And the node `Model` handles permissions. So if your node is private, it will return `null` and the execution can continue. Your private node won't show unless you have permissions to see it.